Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h

```diff
@@ -0,0 +1,101 @@
+
+ public:
+  MeanVertexCalibrator(int minEnt = 500, bool useFit = false, int nBinsX = 100, float rangeX = 1.f,
+                       int nBinsY = 100, float rangeY = 1.f, int nBinsZ = 100, float rangeZ = 20.f,
+                       int nSlotsSMA = 5) : mMinEntries(minEnt), mUseFit(useFit), mNBinsX(nBinsX), mRangeX(rangeX), mNBinsY(nBinsY), mRangeY(rangeY), mNBinsZ(nBinsZ), mRangeZ(rangeZ), mSMAslots(nSlotsSMA)
+  {
+    mSMAdata.init(useFit, nBinsX, rangeX, nBinsY, rangeY, nBinsZ, rangeZ);
+  }
+
+  ~MeanVertexCalibrator() final = default;
+
+  bool hasEnoughData(const Slot& slot) const final
+  {
+    LOG(INFO) << "container entries = " << slot.getContainer()->entries << ", minEntries = " << mMinEntries;
+    return slot.getContainer()->entries >= mMinEntries;
+  }
+  void initOutput() final;
+  void finalizeSlot(Slot& slot) final;
+  Slot& emplaceNewSlot(bool front, TFType tstart, TFType tend) final;
+
+  uint64_t getNSlotsSMA() const { return mSMAslots; }
+  void setNSlotsSMA(uint64_t nslots) { mSMAslots = nslots; }
+
+  void doSimpleMovingAverage(std::deque<float>& dq, float& sma);
+  void doSimpleMovingAverage(std::deque<MVObject>& dq, MVObject& sma);
+
+  const MVObjectVector& getMeanVertexObjectVector() const { return mMeanVertexVector; }
+  const CcdbObjectInfoVector& getMeanVertexObjectInfoVector() const { return mInfoVector; }
+  CcdbObjectInfoVector& getMeanVertexObjectInfoVector() { return mInfoVector; }
+
+ private:
+  int mMinEntries = 0;
+  int mNBinsX = 0;
+  float mRangeX = 0.;
+  int mNBinsY = 0;
+  float mRangeY = 0.;
+  int mNBinsZ = 0;
+  float mRangeZ = 0.;
+  bool mUseFit = false;
+  uint64_t mSMAslots = 5;
+  CcdbObjectInfoVector mInfoVector;        // vector of CCDB Infos , each element is filled with the CCDB description
+                                           // of the accompanying LHCPhase
+  MVObjectVector mMeanVertexVector;        // vector of Mean Vertex Objects, each element is filled in "process"
+                                           // when we finalize one slot (multiple can be finalized during the same
+                                           // "process", which is why we have a vector. Each element is to be considered
+                                           // the output of the device, and will go to the CCDB. It is the simple
+                                           // moving average
+  std::deque<MVObject> mTmpMVobjDq;        // This is the deque of MeanVertex objecs that will be used for the
+                                           // simple moving average
+  MVObject mSMAMVobj;                      // object containing the Simple Moving Average to be put to CCDB
+  std::deque<TFType> mTmpMVobjDqTimeStart; // This is the deque of MeanVertex objecs that will be used for the
+                                           // simple moving average, start time of used TFs
+  std::deque<TFType> mTmpMVobjDqTimeEnd;   // This is the deque of MeanVertex objecs that will be used for the
```

Comment:
Actually, better use std::deque<o2::math_utils::Bracket<TFType>>, to avoid 2 reallocations.
But I think your 1st idea is better: you can define the time range in the MeanVertexData itself.
I did not understand your arguments why this will not work.

Reply:
The point is that mTmpMVobjDqTimeStart is used to define the start validity of the CCDB object. If I bound it to the MeanVertexData itself, when I do the SMA of the fits, I will not have this available, since the case "mUseFit" fills a deque of MeanVertexObjects (no data) on which then I will then do the SMA. 
BTW, I think that the way the `startValidity` is defined is wrong. Currently it just does the average of all start times of the various slots, but this will not necessarily be a uint64_t, as required as start validity for a CCDB entry (now we use the TF id, in the future it should become the timestamp).
In addition, when I have enough slots, I would do a SMA also for the start validity. 
The other possibility is to associate as start validity always the start of the first slot used for the average, but this is not what we discussed. 
What do you think?

Reply:
OK, I see. For the startValidity, indeed, I did not notice you do this in this way. But in any case, it is uint64_t simply because you convert the mean of the all TimeStarts to that. But I think it is much simpler to simply assign the value to the center of the window, i.e. (mTmpMVobjDqTimeStart.front() + mTmpMVobjDqTimeEnd.back())/2;

I did not understand what do you mean by doing SMA for the startValidity? This is a point in time to which you assign the SMA of vertex, why do you need to apply SMA to it?

Reply:
Hi Ruben, 
That is what I wanted to do, assign in the middle of the window. But I forgot to remove the first element of the deque (for the start time), so I was always taking all start times that had been seen. This is the first problem which I need to fix (and indeed I need the end time too). 
Secondly, since I use uint64_t, dividing by 2 will lead to a sort of rounding effect. If this is not an issue, but a good enough approximation, then fine with me. The other possibility was to do the average of the start times of the slots over which I do the running average - this is what I meant. 
I will implement now as you suggested.
Chiara

Reply:
Hi @chiarazampolli 
small rounding is not a problem, the coarsest unit of this uint64_t will be TF (might be eventually an orbit), so the error will be ~5ms, which is irrelevant for the MV variation.

Reply:
Yes, I agree.
Now I fixed the determination of the startValidity. I think nothing else is missing. Right? Aprt for more testing, with more events and TFs.

Reply:
It's ok, but I see you are still using 2 separate dequeus for start and end, leading to 2 reallocations. Why not using a single container with pair of timestamps? The Bracket class I mentioned provides also some comparison operations.

Reply:
Done.